core/services/ocr2/plugins/ocr2keeper/evmregistry/v21/logprovider/buffer_v1.go

```diff
@@ -0,0 +1,420 @@
+		blockThreshold = 1
+	}
+	return buf.enqueue(blockThreshold, logs...)
+}
+
+// Dequeue greedly pulls logs from the buffers.
+// Returns logs and the number of remaining logs in the buffer.
+func (b *logBuffer) Dequeue(block int64, blockRate, upkeepLimit, maxResults int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	start, end := getBlockWindow(block, blockRate)
+	return b.dequeue(start, end, upkeepLimit, maxResults, upkeepSelector)
+}
+
+// dequeue pulls logs from the buffers, depends the given selector (upkeepSelector),
+// in block range [start,end] with minimum number of results per upkeep (upkeepLimit)
+// and the maximum number of results (capacity).
+// Returns logs and the number of remaining logs in the buffer for the given range and selector.
+// NOTE: this method is not thread safe and should be called within a lock.
+func (b *logBuffer) dequeue(start, end int64, upkeepLimit, capacity int, upkeepSelector func(id *big.Int) bool) ([]BufferedLog, int) {
+	var result []BufferedLog
+	var remainingLogs int
+	for _, q := range b.queues {
+		if !upkeepSelector(q.id) {
+			// if the upkeep is not selected, skip it
+			continue
+		}
+		logsInRange := q.sizeOfRange(start, end)
+		if logsInRange == 0 {
+			// if there are no logs in the range, skip the upkeep
+			continue
+		}
+		if capacity == 0 {
+			// if there is no more capacity for results, just count the remaining logs
+			remainingLogs += logsInRange
+			continue
+		}
+		if upkeepLimit > capacity {
+			// adjust limit if it is higher than the actual capacity
+			upkeepLimit = capacity
+		}
+		logs, remaining := q.dequeue(start, end, upkeepLimit)
+		for _, l := range logs {
+			result = append(result, BufferedLog{ID: q.id, Log: l})
+			capacity--
+		}
+		remainingLogs += remaining
+	}
+	return result, remainingLogs
+}
+
+func (b *logBuffer) SetConfig(lookback, blockRate, logLimit uint32) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.opts.override(lookback, blockRate, logLimit)
+}
+
+func (b *logBuffer) NumOfUpkeeps() int {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	return len(b.queues)
+}
+
+func (b *logBuffer) SyncFilters(filterStore UpkeepFilterStore) error {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	for upkeepID := range b.queues {
+		uid := new(big.Int)
+		_, ok := uid.SetString(upkeepID, 10)
+		if ok && !filterStore.Has(uid) {
+			// remove upkeep that is not in the filter store
+			delete(b.queues, upkeepID)
+		}
+	}
+
+	return nil
+}
+
+func (b *logBuffer) getUpkeepQueue(uid *big.Int) (*upkeepLogQueue, bool) {
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	ub, ok := b.queues[uid.String()]
+	return ub, ok
+}
+
+func (b *logBuffer) setUpkeepQueue(uid *big.Int, buf *upkeepLogQueue) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.queues[uid.String()] = buf
+}
+
+// TODO: separate files
+
+type logTriggerState uint8
+
+const (
+	logTriggerStateUnknown logTriggerState = iota
+	logTriggerStateSeen
+	logTriggerStateVisited
```

Comment:
Could we have a comment here explaining what seen and visited mean in the context of log triggers?

Reply:
Just a thought, but it looks like we mark as "seen" when we enqueue, and mark as "visited" when we dequeue; does it make sense to do away with "seen" and "visited" altogether and just have `logTriggerStateEnqueued` and `logTriggerStateDequeued`?